src/SafeMessageHandlers/ClusterManagerWorkflow.workflow.cs

```diff
@@ -0,0 +1,209 @@
+        maxHistoryLength = input.TestContinueAsNew ? 120 : int.MaxValue;
+        sleepInterval = TimeSpan.FromSeconds(input.TestContinueAsNew ? 1 : 600);
+    }
+
+    [WorkflowQuery]
+    public State CurrentState { get; init; }
+
+    [WorkflowRun]
+    public async Task<Result> RunAsync(Input input)
+    {
+        await Workflow.WaitConditionAsync(() => CurrentState.ClusterStarted);
+
+        // Perform health checks at intervals
+        do
+        {
+            await PerformHealthChecksAsync();
+            await Workflow.WaitConditionAsync(
+                () => CurrentState.ClusterShutdown || ShouldContinueAsNew,
+                sleepInterval);
+
+            // Continue as new if needed
+            if (ShouldContinueAsNew)
+            {
+                Workflow.Logger.LogInformation("Continuing as new");
+                throw Workflow.CreateContinueAsNewException((ClusterManagerWorkflow wf) => wf.RunAsync(new()
+                {
+                    State = CurrentState,
+                    TestContinueAsNew = input.TestContinueAsNew,
+                }));
+            }
+        }
+        while (!CurrentState.ClusterShutdown);
+        return new(CurrentState.MaxAssignedNodes, NumAssignedNodes);
+    }
+
+    [WorkflowSignal]
+    public async Task StartClusterAsync()
+    {
+        CurrentState.ClusterStarted = true;
+        foreach (var node in Enumerable.Range(0, 25))
+        {
+            CurrentState.Nodes[$"{node}"] = null;
+        }
+        Workflow.Logger.LogInformation("Cluster started");
+    }
+
+    [WorkflowSignal]
+    public async Task ShutdownClusterAsync()
+    {
+        await Workflow.WaitConditionAsync(() => CurrentState.ClusterStarted);
+        CurrentState.ClusterShutdown = true;
+        Workflow.Logger.LogInformation("Cluster shut down");
+    }
+
+    public record AllocateNodesToJobInput(int NumNodes, string JobName);
+
+    [WorkflowUpdate]
+    public async Task<List<string>> AllocateNodesToJobAsync(AllocateNodesToJobInput input)
+    {
+        await Workflow.WaitConditionAsync(() => CurrentState.ClusterStarted);
+        if (CurrentState.ClusterShutdown)
+        {
+            throw new ApplicationFailureException(
+                "Cannot allocate nodes to a job, cluster is already shut down");
+        }
+        await nodesLock.WaitAsync();
+        try
+        {
+            var unassignedNodes = CurrentState.Nodes.
+                Where(kvp => kvp.Value == null).
+                Select(kvp => kvp.Key).
+                ToList();
+            if (unassignedNodes.Count < input.NumNodes)
+            {
+                throw new ApplicationFailureException(
+                    $"Cannot allocate {input.NumNodes} nodes, have only {unassignedNodes.Count} available");
+            }
+            var assignedNodes = unassignedNodes[..input.NumNodes];
+            // This await would be dangerous without nodesLock because it yields control and allows
+            // interleaving
```

Comment:
```suggestion
            // This await would be dangerous without nodesLock because it yields control and would
            // allow execution to interleave with other concurrent updates
```